### PR TITLE
[julia]: Rewrite to use newly-created tarballs

### PR DESCRIPTION
**Pull request by staticfloat**

### CODE REVIEW

File: J/Julia/build_tarballs.jl [hunk 1]
```diff
@@ -1,103 +1,54 @@
...
-#    "https://github.com/staticfloat/LLVMBuilder/releases/download/v6.0.1-4%2Bnowasm/build_LLVM.v6.0.1.jl",
-#    "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LibGit2-v1.8.0-0/build_LibGit2.v0.27.7.jl",
-#    "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LibSSH2-v1.8.0-0/build_LibSSH2.v1.8.0.jl",
-#    "https://github.com/JuliaWeb/MbedTLSBuilder/releases/download/v0.16.0/build_MbedTLS.v2.13.1.jl",
-    "https://github.com/bicycle1885/ZlibBuilder/releases/download/v1.0.3/build_Zlib.v1.2.11.jl",
-#    "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LibCURL-v7.61.0-0/build_LibCURL.v7.61.0.jl",
-#    "https://github.com/JuliaPackaging/Yggdrasil/releases/download/PCRE2-v10.32-0/build_PCRE2.v10.32.0.jl",
-#    "https://github.com/JuliaMath/GMPBuilder/releases/download/v6.1.2-2/build_GMP.v6.1.2.jl",
-#    "https://github.com/JuliaMath/MPFRBuilder/releases/download/v4.0.1-3/build_MPFR.v4.0.1.jl",
+    Dependency("Zlib"),
```
Comment (giordano) [inline J/Julia/build_tarballs.jl:51]:
```suggestion
    Dependency("Zlib_jll"),
```

File: J/Julia/build_tarballs.jl [hunk 2]
```diff
@@ -1,103 +1,54 @@
...
+
+sources = [ArchiveSource(tarball_url(p), triplet(p)) for p in platforms]
+
+script = raw"""
+cp -rva ${WORKSPACE}/srcdir/${target}/* ${prefix}/
+"""
+
 # The products that we will ensure are always built
-products(prefix) = Product[
+products = Product[
```
Comment (giordano) [inline J/Julia/build_tarballs.jl:44]:
```suggestion
products = [
```

File: J/Julia/build_tarballs.jl [hunk 3]
```diff
@@ -1,103 +1,54 @@
...
     ExecutableProduct(prefix, "julia", :julia),
     LibraryProduct(prefix, "libjulia", :libjulia),
```
Comment (giordano) [inline J/Julia/build_tarballs.jl:46]:
```suggestion
    ExecutableProduct("julia", :julia),
    LibraryProduct("libjulia", :libjulia),
```

File: J/Julia/build_tarballs.jl [hunk 4]
```diff
@@ -1,103 +1,54 @@
...
+    os_tag(p::Windows) = "win$(wordsize(p))"
+    os_tag(p::MacOS) = "mac$(wordsize(p))"
+    function os_tag(p::Linux)
+        if arch(p) in (:armv7l, :aarch64, :ppc64le)
+            return "linux$(arch(p))"
+        end
+        return "linux$(wordsize(p))"
+    end
+    return "https://julialang2.s3.amazonaws.com/bin/$(os_folder(p))/x64/$(version.major).$(version.minor)/julia-$(version)-$(os_tag(p)).tar.gz"
+end
```
Comment (giordano) [inline J/Julia/build_tarballs.jl:35]: Something like
```suggestion
function tarball_url(p::Platform)
    os_folder(p::Windows) = "winnt"
    os_folder(p::MacOS) = "mac"
    os_folder(p::Linux) = "linux"
    os_folder(p::FreeBSD) = "freebsd"

    arch_folder(p) = Dict{Symbol, String}(:i686    => "x86",
                                          :x86_64  => "x64",
                                          :aarch64 => "aarch64",
                                          :armv7l  => "armv7l",
                                          )[arch(p)]
    
    os_tag(p::Windows) = "win$(wordsize(p))"
    os_tag(p::MacOS) = "mac$(wordsize(p))"
    function os_tag(p::Linux)
        if arch(p) in (:armv7l, :aarch64, :ppc64le)
            return "linux-$(arch(p))"
        end
        return "linux-$(wordsize(p))"
    end
    return "https://julialang2.s3.amazonaws.com/bin/$(os_folder(p))/$(arch_folder(p))/$(version.major).$(version.minor)/julia-$(version)-$(os_tag(p)).tar.gz"
end
```
?  I don't know if there is an easier way to get the architecture folder

Comment (staticfloat) [reply J/Julia/build_tarballs.jl]: I've been tweaking this code for too long.... I still have it stuck in my head that it's called `linux64.tar.gz`, not `linux-x86_64.tar.gz`

File: J/Julia/build_tarballs.jl [hunk 5]
```diff
@@ -1,103 +1,50 @@
...
-#BUILD_FLAGS+=(LDFLAGS="$LDFLAGS")
-BUILD_FLAGS+=(LIBSSH2_ENABLE_TESTS=0)
-
-make ${BUILD_FLAGS[@]} -j${nproc} -C deps install-mbedtls
-make ${BUILD_FLAGS[@]} -j${nproc} -C deps install-libgit2
-make ${BUILD_FLAGS[@]} -j${nproc}
-make ${BUILD_FLAGS[@]} install
-"""
+version = v"1.4.1"
+verbose = "--verbose" in ARGS
```
Comment (giordano) [inline J/Julia/build_tarballs.jl:5]: I don't think this is used anywhere?
```suggestion
```

**Comment by staticfloat** [comment]
Windows errors will be fixed by https://github.com/staticfloat/ObjectFile.jl/pull/23

File: J/Julia/build_tarballs.jl [hunk 6]
```diff
@@ -1,103 +1,50 @@
...
-    Linux(:i686; libc=:glibc, compiler_abi=CompilerABI(:gcc7)),
-    Windows(:x86_64; compiler_abi=CompilerABI(:gcc7)),
-    Windows(:i686; compiler_abi=CompilerABI(:gcc7)),
+    Linux(:armv7l; libc=:glibc, compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    Linux(:aarch64; libc=:glibc, compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    Linux(:x86_64; libc=:glibc, compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    Linux(:i686; libc=:glibc, compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    MacOS(:x86_64; compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    Windows(:x86_64; compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
+    Windows(:i686; compiler_abi=CompilerABI(libgfortran_version=v"4", cxxstring_abi=:cxx11)),
```
Comment (ararslan) [inline J/Julia/build_tarballs.jl:15]: FreeBSD too? :wink:

Comment (ararslan) [reply J/Julia/build_tarballs.jl:15]: Looks like it was added as a source but not a platform.

Comment (giordano) [reply J/Julia/build_tarballs.jl:15]: Oh no!  [This commit message](https://github.com/JuliaPackaging/Yggdrasil/pull/1048/commits/4cd3ba16336a98214eb0b2e39b9ab1c17c9b6712) sounds more sarcastic now :sweat_smile:

**Pull request merged**